src/main/java/org/spongepowered/common/event/tracking/phase/tick/EntityTickPhaseState.java

```diff
@@ -148,9 +152,23 @@ protected void processCaptures(final Entity tickingEntity, final EntityTickConte
                     }
                     if (!projectile.isEmpty()) {
                         frame.addContext(EventContextKeys.SPAWN_TYPE, SpawnTypes.PROJECTILE);
-                        SpongeCommonEventFactory.callSpawnEntity(projectile, phaseContext);
-                        frame.removeContext(EventContextKeys.SPAWN_TYPE);
+                        if (tickingEntity instanceof ProjectileSource) {
+                            frame.addContext(EventContextKeys.PROJECTILE_SOURCE, (ProjectileSource) tickingEntity);
+                        }
+                        final Cause cause = Sponge.getCauseStackManager().getCurrentCause();
+                        boolean launchEventCancelled = false;
+                        for (Entity proj : projectile) {
+                            if (SpongeImpl.postEvent(SpongeEventFactory.createLaunchProjectileEvent(cause, (Projectile) proj))) {
+                                launchEventCancelled = true;
```

Comment:
So if any of the launch events are cancelled, then how would you tell plugins that the events prior were cancelled?